src/cloudflare/internal/d1-api.ts

```diff
@@ -52,38 +52,96 @@ type SQLError = {
 
+type D1SessionCommitTokenOrConstraint = string;
+type D1SessionCommitToken = string;
+
+// TODO Figure out what header to use.
+const D1_SESSION_COMMIT_TOKEN_HTTP_HEADER = 'x-cf-d1-session-commit-token';
+
 class D1Database {
-  private readonly fetcher: Fetcher;
+  private readonly envFetcher: Fetcher;
+  private readonly alwaysPrimarySession: D1DatabaseSessionAlwaysPrimary;
 
-  public constructor(fetcher: Fetcher) {
-    this.fetcher = fetcher;
+  public constructor(envFetcher: Fetcher) {
+    this.envFetcher = envFetcher;
+    this.alwaysPrimarySession = new D1DatabaseSessionAlwaysPrimary(
+      this.envFetcher
+    );
+  }
+
+  public withSession(
+    constraintOrToken: D1SessionCommitTokenOrConstraint | null | undefined
+  ): D1DatabaseSession {
+    constraintOrToken = constraintOrToken?.trim();
+    if (
+      constraintOrToken === null ||
+      constraintOrToken === undefined ||
+      constraintOrToken === ''
+    ) {
+      constraintOrToken = 'first-unconstrained';
+    }
+    return new D1DatabaseSession(this.envFetcher, constraintOrToken);
   }
 
   public prepare(query: string): D1PreparedStatement {
-    return new D1PreparedStatement(this, query);
+    return new D1PreparedStatement(this.alwaysPrimarySession, query);
+  }
+
+  public async batch<T = unknown>(
+    statements: D1PreparedStatement[]
+  ): Promise<D1Result<T>[]> {
+    return this.alwaysPrimarySession.batch(statements);
+  }
+
+  public async exec(query: string): Promise<D1ExecResult> {
+    return this.alwaysPrimarySession.exec(query);
   }
 
   // DEPRECATED, TO BE REMOVED WITH NEXT BREAKING CHANGE
   public async dump(): Promise<ArrayBuffer> {
-    const response = await this.fetcher.fetch('http://d1/dump', {
-      method: 'POST',
-      headers: {
-        'content-type': 'application/json',
-      },
-    });
-    if (response.status !== 200) {
-      try {
-        const err = (await response.json()) as SQLError;
-        throw new Error(`D1_DUMP_ERROR: ${err.error}`, {
-          cause: new Error(err.error),
-        });
-      } catch {
-        throw new Error(`D1_DUMP_ERROR: Status + ${response.status}`, {
-          cause: new Error(`Status ${response.status}`),
-        });
-      }
+    return this.alwaysPrimarySession.dump();
+  }
+}
+
+class D1DatabaseSession {
+  protected fetcher: Fetcher;
+  protected commitTokenOrConstraint: D1SessionCommitTokenOrConstraint;
+
+  public constructor(
+    envFetcher: Fetcher,
+    commitTokenOrConstraint: D1SessionCommitTokenOrConstraint
+  ) {
+    this.fetcher = envFetcher;
+    this.commitTokenOrConstraint = commitTokenOrConstraint;
+
+    // TODO Do format validation for the commit token?
+    if (!this.commitTokenOrConstraint) {
+      throw new Error('D1_SESSION_ERROR: invalid commit token or constraint');
     }
-    return await response.arrayBuffer();
+  }
+
+  // Update the commit token IFF the given newCommitToken is more recent.
```

Comment:
I take it that this is the part of the code that protects against the case where there are multiple requests in flight from the same session.  It might be worth dropping a comment saying so!

Reply:
Yeap, that's its purpose. I will add a comment further explaining it.